# Computer Class 1a - pandas and data aggregation

The examples and exercises of this computer class introduce the student to working with the pandas library. It can be used in conjunction with chapters 5-8 and 10 of the McKinney book.

*Authors: Cees Diks and Bram Wouters, Faculty Economics and Business, University of Amsterdam (UvA)* <br>
*Copyright (C): UvA (2023)* <br>
*Credits: some of the examples and formulations are taken from McKinney.*

In [1]:
# Importing the relevant modules
import numpy as np
import pandas as pd

# 1. Creating a Series and Series basics

**Exercise 1:** create a Series with indices 'United States', 'Netherlands', 'Australia','Uruguay' and values 'North America', 'Europe', 'Oceania', 'South America', respectively. Call the Series `ser1` and show the result.

In [5]:
ser1 = pd.Series(['North America', 'Europe', 'Oceania', 'South America'], index=["United States", 'Netherlands', 'Australia','Uruguay'])
ser1

United States    North America
Netherlands             Europe
Australia              Oceania
Uruguay          South America
dtype: object

**Exercise 2:** use a dictionary to create a second Series (called `ser2`) with indices 'Netherlands','China','France' and values 'Europe', 'Asia', respectively. The value corresponding to the index 'France' is missing on purpose and Python reports it as NaN (Not a Number). Show the result.

In [4]:
d = {"Netherlands": "Europe", "China": "Asia"}
ser2 = pd.Series(d, index=["Netherlands", "China", "France"])
ser2

Netherlands    Europe
China            Asia
France            NaN
dtype: object

**Exercise 3:** use the pandas function `concat` to merge `ser1` and `ser2`. Call the resulting Series `ser3` and print the result.

In [7]:
ser3 = pd.concat([ser1, ser2])
ser3

United States    North America
Netherlands             Europe
Australia              Oceania
Uruguay          South America
Netherlands             Europe
China                     Asia
France                     NaN
dtype: object

**Exercise 4:** remove the duplicated data from `ser3`. Note that instead of redefining `ser3`, you can also use the `inplace` keyword. Print the resulting `ser3`. (Hint: use the Series method `drop_duplicates`.)

In [8]:
ser3.drop_duplicates(inplace=True)
ser3

United States    North America
Netherlands             Europe
Australia              Oceania
Uruguay          South America
China                     Asia
France                     NaN
dtype: object

**Exercise 5:** create a boolean that checks whether there are any missing values (NaN) in `ser3`. 

In [9]:
ser3.isna()

United States    False
Netherlands      False
Australia        False
Uruguay          False
China            False
France            True
dtype: bool

**Exercise 6:** replace the missing value in `ser3` by the correct continent.

In [10]:
ser3.fillna("Europe", inplace=True)

**Exercise 7:** a Series and the index of a Series both have a name attribute. This can be useful when dealing with many different Series and/or when multiple people need to work with the same Series. Give `ser3` the following name: 'continent'. Also, give the index of `ser3` the name 'country' and print the resulting `ser3`.

In [11]:
ser3.name = "Continents"
ser3.index.name = "Countries"
ser3

Countries
United States    North America
Netherlands             Europe
Australia              Oceania
Uruguay          South America
China                     Asia
France                  Europe
Name: Continents, dtype: object

# 2. Creating a DataFrame and DataFrame basics

**Exercise 8:** create DataFrame that corresponds to the table below and bind the variable `df1` to it. The first column in the table below should serve as index of the DataFrame. Make sure the column indices match the column names of the table. Show the resulting `df1` as output.

|**country**|**GDP (nominal, \$MM)**| **population (MM)**|
|:-------|:-------------|:--------|
| **United States**        |    20412870          |    327.4     |
| **France**        |    2925096          |    67.2     |
| **Netherlands**        |   945327          |    17.2     |
| **Australia**        |    1500256          |    25.0     |
| **Uruguay**        |    63370          |    3.5     |

In [14]:
d = {"United States": [20412870, 327.4], "France": [2925096, 67.2], "Netherlands": [945327, 17.2], "Australia": [1500256, 25.0], "Uruguay": [63370, 3.5]}
df1 = pd.DataFrame(d, index=["GDP (nominal, $MM)", "population (MM)"]).T
df1.index.name = "country"
df1

GDP (nominal, $MM)  population (MM)
country                                           
United States          20412870.0            327.4
France                  2925096.0             67.2
Netherlands              945327.0             17.2
Australia               1500256.0             25.0
Uruguay                   63370.0              3.5

**Exercise 9:** use square brackets `[]` to extract the population column from the DataFrame and print it. Also, verify explicitely that this column is a Series object.

In [15]:
pop = df1["population (MM)"]
assert isinstance(pop, pd.Series)
pop

country
United States    327.4
France            67.2
Netherlands       17.2
Australia         25.0
Uruguay            3.5
Name: population (MM), dtype: float64

**Exercise 10:** add `ser3` as an extra column to `df1` and call the new DataFrame `df2`. Make sure that China is included as a row in `df2`. (Hint: turn `ser3` into a DataFrame with the method `to_frame` and subsequently use `merge`. Use the [documentation](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.merge.html) to find out which keywords you need.)

In [24]:
df2 = df1.merge(ser3.to_frame(), right_index=True, left_index=True, how="outer")
df2

GDP (nominal, $MM)  population (MM)     Continents
Australia               1500256.0             25.0        Oceania
China                         NaN              NaN           Asia
France                  2925096.0             67.2         Europe
Netherlands              945327.0             17.2         Europe
United States          20412870.0            327.4  North America
Uruguay                   63370.0              3.5  South America

**Exercise 11:** replace the missing values for China by the actual values. (Note: use `at[...]` for accessing values of a DataFrame. Don't use `get_value`; this is an old method that will be discarded in future versions of pandas.) Also note that for indexing one can use both index/column names and integers.

In [26]:
df2.at["China", "GDP (nominal, $MM)"] = 17960000
df2.at["China", "population (MM)"] = 1409.7
df2

GDP (nominal, $MM)  population (MM)     Continents
Australia               1500256.0             25.0        Oceania
China                  17960000.0           1409.7           Asia
France                  2925096.0             67.2         Europe
Netherlands              945327.0             17.2         Europe
United States          20412870.0            327.4  North America
Uruguay                   63370.0              3.5  South America

**Exercise 12:** add another column called 'visited?' to `df2`. The values must be booleans and indicate whether or not you have visited the country.

In [27]:
df2["visited"] = [False, False, True, True, True, False]
df2

GDP (nominal, $MM)  population (MM)     Continents  visited
Australia               1500256.0             25.0        Oceania    False
China                  17960000.0           1409.7           Asia    False
France                  2925096.0             67.2         Europe     True
Netherlands              945327.0             17.2         Europe     True
United States          20412870.0            327.4  North America     True
Uruguay                   63370.0              3.5  South America    False

**Exercise 13:** add another column called 'GDP per capita (nominal, \$)' to `df2`. The values are the nominal GDP per capita, which can be calculated with the data in the other columns of the DataFrame.

In [29]:
df2["GDP per capita (nominal, $)"] = df2["GDP (nominal, $MM)"] / df2["population (MM)"]
df2

GDP (nominal, $MM)  population (MM)     Continents  visited  \
Australia               1500256.0             25.0        Oceania    False   
China                  17960000.0           1409.7           Asia    False   
France                  2925096.0             67.2         Europe     True   
Netherlands              945327.0             17.2         Europe     True   
United States          20412870.0            327.4  North America     True   
Uruguay                   63370.0              3.5  South America    False   

               GDP per capita (nominal, $)  
Australia                     60010.240000  
China                         12740.299354  
France                        43528.214286  
Netherlands                   54960.872093  
United States                 62348.411729  
Uruguay                       18105.714286

**Exercise 14:** sort the rows of `df2` by GDP per capita in descending order and show the resulting `df2` as output. (Hint: choose between `sort_index` and `sort_values` which one to use.)

In [30]:
df2.sort_values(by="GDP per capita (nominal, $)", ascending=False)

GDP (nominal, $MM)  population (MM)     Continents  visited  \
United States          20412870.0            327.4  North America     True   
Australia               1500256.0             25.0        Oceania    False   
Netherlands              945327.0             17.2         Europe     True   
France                  2925096.0             67.2         Europe     True   
Uruguay                   63370.0              3.5  South America    False   
China                  17960000.0           1409.7           Asia    False   

               GDP per capita (nominal, $)  
United States                 62348.411729  
Australia                     60010.240000  
Netherlands                   54960.872093  
France                        43528.214286  
Uruguay                       18105.714286  
China                         12740.299354

## Slicing and filtering

There exist two methods  for slicing a DataFrame: `loc` when using index/column names and `iloc` when using integers to specify rows/columns. Note that slicing with `loc` is inclusive with respect to the last element, while slicing with `iloc` is exclusive (as was the case for NumPy's ndarrays).

**Exercise 15:** use `loc` to extract the row about the Netherlands from `df2`. (Note that in this case a Series is returned.)

In [31]:
df2.loc["Netherlands"]

GDP (nominal, $MM)                 945327.0
population (MM)                        17.2
Continents                           Europe
visited                                True
GDP per capita (nominal, $)    54960.872093
Name: Netherlands, dtype: object

**Exercise 16:** use `loc` to extract the rows up to and including the Netherlands from `df2`. (Note that in this case a DataFrame is returned.)

In [32]:
df2.loc[:"Netherlands"]

GDP (nominal, $MM)  population (MM) Continents  visited  \
Australia             1500256.0             25.0    Oceania    False   
China                17960000.0           1409.7       Asia    False   
France                2925096.0             67.2     Europe     True   
Netherlands            945327.0             17.2     Europe     True   

             GDP per capita (nominal, $)  
Australia                   60010.240000  
China                       12740.299354  
France                      43528.214286  
Netherlands                 54960.872093

**Exercise 17:** extract the rows up to and including the Netherlands from `df2`, and only the columns about the population and GDP per capita.

In [33]:
df2.loc[:"Netherlands"]["GDP per capita (nominal, $)"]

Australia      60010.240000
China          12740.299354
France         43528.214286
Netherlands    54960.872093
Name: GDP per capita (nominal, $), dtype: float64

**Exercise 18:** if you have used `loc` in the previous exercise, now use `iloc` to repeat it (and vice versa).

In [34]:
df2.iloc[:4]

GDP (nominal, $MM)  population (MM) Continents  visited  \
Australia             1500256.0             25.0    Oceania    False   
China                17960000.0           1409.7       Asia    False   
France                2925096.0             67.2     Europe     True   
Netherlands            945327.0             17.2     Europe     True   

             GDP per capita (nominal, $)  
Australia                   60010.240000  
China                       12740.299354  
France                      43528.214286  
Netherlands                 54960.872093

In [35]:
df2.iloc[:4]["GDP per capita (nominal, $)"]

Australia      60010.240000
China          12740.299354
France         43528.214286
Netherlands    54960.872093
Name: GDP per capita (nominal, $), dtype: float64

**Exercise 19:** filtering rows from a DataFrame proceeds in the same fashion as for NumPy's ndarray, i.e. by using broadcasting with a Series of booleans. Create a DataFrame including only those countries in `df2` whose population is above 50 million.

In [36]:
df2[df2["population (MM)"] > 50]

GDP (nominal, $MM)  population (MM)     Continents  visited  \
China                  17960000.0           1409.7           Asia    False   
France                  2925096.0             67.2         Europe     True   
United States          20412870.0            327.4  North America     True   

               GDP per capita (nominal, $)  
China                         12740.299354  
France                        43528.214286  
United States                 62348.411729

**Exercise 20:** create a DataFrame with only the countries you have visited.

In [38]:
df2[df2["visited"]]

GDP (nominal, $MM)  population (MM)     Continents  visited  \
France                  2925096.0             67.2         Europe     True   
Netherlands              945327.0             17.2         Europe     True   
United States          20412870.0            327.4  North America     True   

               GDP per capita (nominal, $)  
France                        43528.214286  
Netherlands                   54960.872093  
United States                 62348.411729

# 3. DataFrame manipulation

In the following exercises we will be working with existing data. The data is stored in an Excel file ('weather_accident_data.xlsx') and can be found on Canvas. The goal is to become acquainted with basic data cleaning and manipulation of data in a pandas DataFrame.

The data consists of daily weather data for the Netherlands for January 2018 and the daily number of traffic accidents during the same period. Parts of the data are fictional.

**Exercise 21:** import the Excel file ('weather_accident_data.xlsx') as a pandas DataFrame. The simplest way to do this is to put the Excel file in the same folder as the Jupyter notebook and the use the `read_excel` function. Inspect the data by printing it. It includes information about: 
* the date
* wind
* average temperature
* minimum temperature
* maximum temperature
* hours of sun
* hours of rain
* amount of rain
* fog
* snow
* issuing of a weather alarm
* the number of traffic accidents

The structure of the Excel file is not very practical. We would like each row to become a column in the pandas DataFrame, so you will need to transpose the data. The first column in the Excel file should become the column names of the DataFrame. The index of the DataFrame should be the default (i.e. integers starting at 0). 

Finally, give the created DataFrame the name `df` and print the first ten rows.

In [47]:
df = pd.read_excel("weather_accident_data.xlsx",header=None).T
df.columns = df.iloc[0]
df = df[1:]
df.head(10)

0   YYYYMMDD       - T_ave T_low T_high  sun rain_duration rain(mm)  fog snow  \
1   20180123  3.2bft   8.1   4.6   11.2  0.9             4      1.5  m/e    4   
2   20180201  2.8bft   3.4   1.9    6.1  2.5           4.3      4.4    m    0   
3   20180131  3.4bft   6.5   3.2    9.5    0           6.4      3.8    m    0   
4   20180114  2.9bft   2.9   0.8    5.2    5             0        0    -    0   
5   20180105  2.8bft   6.4   4.4    8.7    0           2.4      2.7    m    0   
6   20180120  2.1bft   2.2   0.8    3.6  0.4           NaN      NaN  m/e    1   
7   20171230  3.5bft   8.8   3.9   12.2    0             5      3.7    -    0   
8   20180103  4.8bft   8.8   5.4   11.2  0.2           5.1      9.5    -    0   
9   20180102  3.0bft   6.5   4.5    9.1  1.4           6.2      4.5    -    0   
10  20180129  3.6bft   8.8   4.6   10.7    0           2.8      5.2    -    0   

0  test_probe weather alarm accidents  
1         ###    red: storm        83  
2         ###           NaN        91  
3         ###           NaN        86  
4         ###           NaN        92  
5         ###           NaN        75  
6         ###           NaN        74  
7         ###           NaN       114  
8         ###           NaN       135  
9         ###           NaN       102  
10        ###           NaN       124

**Exercise 22:** the name of the wind column is '-' instead of 'wind'. Give the column the correct name.

In [49]:
df.rename(columns={"-": "wind"}, inplace=True)

**Exercise 23:** sort the rows of the DataFrame by date.

In [51]:
df.sort_values(by="YYYYMMDD", inplace=True)
df.head()

0   YYYYMMDD    wind T_ave T_low T_high  sun rain_duration rain(mm) fog snow  \
27  20171229  3.2bft   2.8  -0.7    6.7  0.3           6.3      6.7   -    0   
7   20171230  3.5bft   8.8   3.9   12.2    0             5      3.7   -    0   
21  20171231  3.8bft  10.6   8.5   13.7  0.2          15.3     39.3   -    0   
11  20180101  3.2bft   6.8   5.2    8.8  2.1           6.9      4.7   -    0   
9   20180102  3.0bft   6.5   4.5    9.1  1.4           6.2      4.5   -    0   

0  test_probe weather alarm accidents  
27        ###           NaN       125  
7         ###           NaN       114  
21        ###           NaN       416  
11        ###           NaN        91  
9         ###           NaN       102

**Exercise 24:** note that the index of `df` is mixed up due to the sorting. Reset the index of `df`.

In [53]:
df.reset_index(inplace=True)

**Exercise 25:** accidentally the data set includes a few days in December and February. Remove those rows from `df` and reset the index again.

In [62]:
df["YYYYMMDD"] = df["YYYYMMDD"].apply(str)
df = df[~df["YYYYMMDD"].str.contains("....[0,1]2..")]
df.reset_index(drop=True, inplace=True)
df

0   index  YYYYMMDD    wind T_ave T_low T_high  sun rain_duration rain(mm)  \
0      11  20180101  3.2bft   6.8   5.2    8.8  2.1           6.9      4.7   
1       9  20180102  3.0bft   6.5   4.5    9.1  1.4           6.2      4.5   
2       8  20180103  4.8bft   8.8   5.4   11.2  0.2           5.1      9.5   
3      29  20180104  3.5bft   8.2   6.8   10.7  0.2           1.6      0.6   
4       5  20180105  2.8bft   6.4   4.4    8.7    0           2.4      2.7   
5      19  20180106  2.2bft   4.5   2.2    6.6  0.7             0        0   
6      33  20180107  3.6bft     1  -1.2    3.1  6.2             0        0   
7      30  20180108  3.5bft   1.1  -1.4    3.9  6.5             0        0   
8      35  20180109  2.4bft     3  -0.2    6.1  0.3             0        0   
9      22  20180110  2.2bft   6.9   5.8    7.6    0           5.8      2.6   
10     24  20180111  1.5bft   5.7   2.8    7.6    0             0        0   
11     20  20180112  1.4bft     5   4.1    5.9    0             0        0   
12     26  20180113  2.3bft   4.2   1.2    6.2    0             0        0   
13      4  20180114  2.9bft   2.9   0.8    5.2    5             0        0   
14     32  20180115  3.9bft   5.2   1.7      8    0           9.5     13.3   
15     13  20180116  3.2bft   5.2   2.5    6.9  2.9           3.9      8.3   
16     28  20180117  3.4bft   4.6     2      6  0.5           2.2      2.7   
17     16  20180118  3.8bft   4.7   0.5   10.2  0.4           NaN      NaN   
18     25  20180119  2.3bft   3.7   1.4    6.9  2.5           3.1      5.4   
19      6  20180120  2.1bft   2.2   0.8    3.6  0.4           NaN      NaN   
20     31  20180121  2.1bft     3  -0.4    6.9  4.8           2.7      2.1   
21     17  20180122  2.5bft   6.3     2    8.4  0.5           1.9      0.6   
22      1  20180123  3.2bft   8.1   4.6   11.2  0.9             4      1.5   
23     12  20180124  4.1bft  11.7   7.3   14.4    0           4.1      4.2   
24     18  20180125  2.7bft   7.7     5    9.2  0.6           0.1      0.1   
25     34  20180126  1.6bft   5.4   2.5    8.6  1.2             0        0   
26     14  20180127  3.2bft   5.8   4.4    7.2    2           3.6      1.1   
27     23  20180128  3.4bft   9.7   6.9   10.8  1.7             0     -0.1   
28     10  20180129  3.6bft   8.8   4.6   10.7    0           2.8      5.2   
29     15  20180130  2.0bft   4.7   2.1    7.4  6.2             0        0   
30      3  20180131  3.4bft   6.5   3.2    9.5    0           6.4      3.8   

0   fog snow test_probe  weather alarm accidents  
0     -    0        ###            NaN        91  
1     -    0        ###            NaN       102  
2     -    0        ###            NaN       135  
3     m    0        ###            NaN        68  
4     m    0        ###            NaN        75  
5     m    0        ###            NaN        53  
6   m/e    0        ###  yellow: icing        69  
7   m/e    0        ###  yellow: icing        68  
8   m/e    0        ###  yellow: icing        63  
9     m    0        ###            NaN        61  
10    m    0        ###            NaN        83  
11    -    0        ###            NaN        60  
12    -    0        ###            NaN        50  
13    -    0        ###            NaN        92  
14    -    0        ###            NaN       217  
15    -    0        ###            NaN       151  
16    -    0        ###            NaN        95  
17    -    0        ###            NaN       186  
18    -    1        ###            NaN       143  
19  m/e    1        ###            NaN        74  
20  m/e    1        ###  yellow: storm       113  
21  m/e    4        ###     red: storm        64  
22  m/e    4        ###     red: storm        83  
23    e    1        ###            NaN       101  
24    e    0        ###            NaN        74  
25  m/e    0        ###            NaN        47  
26  m/e    0        ###            NaN        68  
27    -    0        ###            NaN        38  
28    -    0        ### 

**Exercise 26:** the column 'test_probe' was created in the data collection process, but does not contain any information. Remove that column from `df`.

In [64]:
df.drop(columns=["test_probe"], inplace=True)

KeyError: "['test_probe'] not found in axis"

**Example:** NumPy's universal functions can also be applied to Series and DataFrames. Note that we use `astype` to create a Series/DataFrame with data of the same type. NumPy's universal functions require this.

In [65]:
print(np.exp(df['rain(mm)'].astype(float)).head(5)) # Applying a universal function to a Series.

print('\n')

print(np.exp(df[['rain(mm)', 'rain_duration']].astype(float)).head(5)) # Applying a universal function to a DataFrame.

np.subtract(df['T_high'], df['T_low']).head(5) # Applying a binary universal function to two Series.

0      109.947172
1       90.017131
2    13359.726830
3        1.822119
4       14.879732
Name: rain(mm), dtype: float64


0      rain(mm)  rain_duration
0    109.947172     992.274716
1     90.017131     492.749041
2  13359.726830     164.021907
3      1.822119       4.953032
4     14.879732      11.023176


0    3.6
1    4.6
2    5.8
3    3.9
4    4.3
dtype: object

**Example:** pandas objects have a number of mathematical and statistical methods. Most of them are reductions (i.e., they extract a single value from a Series or (columns/rows of) a DataFrame) and are similar to the NumPy methods. An important difference is that the pandas methods can handle missing values by ignoring them. For example, in the cell below we compute the standard deviation on a NumPy array and a Series, both containing missing values.

In [66]:
print(df['rain_duration'].values.std()) # Applying a reduction to an ndarray. The values attribute returns a NumPy array.

print(df['rain_duration'].std()) # Applying a reduction to a Series.

nan
2.6669847921901244


**Exercise 27:** the data in the 'rain(mm)' column needs some cleaning:
* one of the entries accidentally has a minus sign. Remove the minus sign.
* replace the missing values by the daily average amount of rain for January. (Hint: use `fillna`.)
* round the values to 1 decimal.

In [71]:
df["rain(mm)"] = df["rain(mm)"].abs()
df["rain(mm)"] = df["rain(mm)"].astype(float)
df.fillna(df["rain(mm)"].mean(), inplace=True)
df["rain(mm)"] = df["rain(mm)"].round(1)
df

0   index  YYYYMMDD    wind  T_ave  T_low  T_high  sun  rain_duration  \
0      11  20180101  3.2bft    6.8    5.2     8.8  2.1       6.900000   
1       9  20180102  3.0bft    6.5    4.5     9.1  1.4       6.200000   
2       8  20180103  4.8bft    8.8    5.4    11.2  0.2       5.100000   
3      29  20180104  3.5bft    8.2    6.8    10.7  0.2       1.600000   
4       5  20180105  2.8bft    6.4    4.4     8.7  0.0       2.400000   
5      19  20180106  2.2bft    4.5    2.2     6.6  0.7       0.000000   
6      33  20180107  3.6bft    1.0   -1.2     3.1  6.2       0.000000   
7      30  20180108  3.5bft    1.1   -1.4     3.9  6.5       0.000000   
8      35  20180109  2.4bft    3.0   -0.2     6.1  0.3       0.000000   
9      22  20180110  2.2bft    6.9    5.8     7.6  0.0       5.800000   
10     24  20180111  1.5bft    5.7    2.8     7.6  0.0       0.000000   
11     20  20180112  1.4bft    5.0    4.1     5.9  0.0       0.000000   
12     26  20180113  2.3bft    4.2    1.2     6.2  0.0       0.000000   
13      4  20180114  2.9bft    2.9    0.8     5.2  5.0       0.000000   
14     32  20180115  3.9bft    5.2    1.7     8.0  0.0       9.500000   
15     13  20180116  3.2bft    5.2    2.5     6.9  2.9       3.900000   
16     28  20180117  3.4bft    4.6    2.0     6.0  0.5       2.200000   
17     16  20180118  3.8bft    4.7    0.5    10.2  0.4       2.517241   
18     25  20180119  2.3bft    3.7    1.4     6.9  2.5       3.100000   
19      6  20180120  2.1bft    2.2    0.8     3.6  0.4       2.517241   
20     31  20180121  2.1bft    3.0   -0.4     6.9  4.8       2.700000   
21     17  20180122  2.5bft    6.3    2.0     8.4  0.5       1.900000   
22      1  20180123  3.2bft    8.1    4.6    11.2  0.9       4.000000   
23     12  20180124  4.1bft   11.7    7.3    14.4  0.0       4.100000   
24     18  20180125  2.7bft    7.7    5.0     9.2  0.6       0.100000   
25     34  20180126  1.6bft    5.4    2.5     8.6  1.2       0.000000   
26     14  20180127  3.2bft    5.8    4.4     7.2  2.0       3.600000   
27     23  20180128  3.4bft    9.7    6.9    10.8  1.7       0.000000   
28     10  20180129  3.6bft    8.8    4.6    10.7  0.0       2.800000   
29     15  20180130  2.0bft    4.7    2.1     7.4  6.2       0.000000   
30      3  20180131  3.4bft    6.5    3.2     9.5  0.0       6.400000   

0   rain(mm)  fog  snow  weather alarm  accidents  
0        4.7    -     0       2.517241         91  
1        4.5    -     0       2.517241        102  
2        9.5    -     0       2.517241        135  
3        0.6    m     0       2.517241         68  
4        2.7    m     0       2.517241         75  
5        0.0    m     0       2.517241         53  
6        0.0  m/e     0  yellow: icing         69  
7        0.0  m/e     0  yellow: icing         68  
8        0.0  m/e     0  yellow: icing         63  
9        2.6    m     0       2.517241         61  
10       0.0    m     0       2.517241         83  
11       0.0    -     0       2.517241         60  
12       0.0    -     0       2.517241         50  
13       0.0    -     0       2.517241         92  
14      13.3    -     0       2.517241        217  
15       8.3    -     0       2.517241        151  
16       2.7    -     0       2.517241         95  
17       2.5    -     0       2.517241        186  
18       5.4    -     1       2.517241        143  
19       2.5  m/e     1       2.517241         74  
20       2.1  m/e     1  yellow: storm        113  
21       0.6  m/e     4     red: storm         64  
22       1.5  m/e     4     red: storm         83  
23       4.2    e     1       2.517241        101  
24       0.1    e     0       2.517241         74  
25       0.0  m/e     0       2.517241         47  
26       1.1  m/e     0       2.517241         68  
27       0.1    -     0       2.517241         38  
28       5.2    -     0       2.517241        124  
29       0.0    m     0       2.517241         77  
30       3.8    m     0       2.517241         86

**Exercise 28:** the values of the 'wind' column are strings. Replace them by numerical types, leaving out the 'bft'.

In [72]:
df["wind"] = df["wind"].str.replace("bft", "").astype(float)
df

0   index  YYYYMMDD  wind  T_ave  T_low  T_high  sun  rain_duration  rain(mm)  \
0      11  20180101   3.2    6.8    5.2     8.8  2.1       6.900000       4.7   
1       9  20180102   3.0    6.5    4.5     9.1  1.4       6.200000       4.5   
2       8  20180103   4.8    8.8    5.4    11.2  0.2       5.100000       9.5   
3      29  20180104   3.5    8.2    6.8    10.7  0.2       1.600000       0.6   
4       5  20180105   2.8    6.4    4.4     8.7  0.0       2.400000       2.7   
5      19  20180106   2.2    4.5    2.2     6.6  0.7       0.000000       0.0   
6      33  20180107   3.6    1.0   -1.2     3.1  6.2       0.000000       0.0   
7      30  20180108   3.5    1.1   -1.4     3.9  6.5       0.000000       0.0   
8      35  20180109   2.4    3.0   -0.2     6.1  0.3       0.000000       0.0   
9      22  20180110   2.2    6.9    5.8     7.6  0.0       5.800000       2.6   
10     24  20180111   1.5    5.7    2.8     7.6  0.0       0.000000       0.0   
11     20  20180112   1.4    5.0    4.1     5.9  0.0       0.000000       0.0   
12     26  20180113   2.3    4.2    1.2     6.2  0.0       0.000000       0.0   
13      4  20180114   2.9    2.9    0.8     5.2  5.0       0.000000       0.0   
14     32  20180115   3.9    5.2    1.7     8.0  0.0       9.500000      13.3   
15     13  20180116   3.2    5.2    2.5     6.9  2.9       3.900000       8.3   
16     28  20180117   3.4    4.6    2.0     6.0  0.5       2.200000       2.7   
17     16  20180118   3.8    4.7    0.5    10.2  0.4       2.517241       2.5   
18     25  20180119   2.3    3.7    1.4     6.9  2.5       3.100000       5.4   
19      6  20180120   2.1    2.2    0.8     3.6  0.4       2.517241       2.5   
20     31  20180121   2.1    3.0   -0.4     6.9  4.8       2.700000       2.1   
21     17  20180122   2.5    6.3    2.0     8.4  0.5       1.900000       0.6   
22      1  20180123   3.2    8.1    4.6    11.2  0.9       4.000000       1.5   
23     12  20180124   4.1   11.7    7.3    14.4  0.0       4.100000       4.2   
24     18  20180125   2.7    7.7    5.0     9.2  0.6       0.100000       0.1   
25     34  20180126   1.6    5.4    2.5     8.6  1.2       0.000000       0.0   
26     14  20180127   3.2    5.8    4.4     7.2  2.0       3.600000       1.1   
27     23  20180128   3.4    9.7    6.9    10.8  1.7       0.000000       0.1   
28     10  20180129   3.6    8.8    4.6    10.7  0.0       2.800000       5.2   
29     15  20180130   2.0    4.7    2.1     7.4  6.2       0.000000       0.0   
30      3  20180131   3.4    6.5    3.2     9.5  0.0       6.400000       3.8   

0   fog  snow  weather alarm  accidents  
0     -     0       2.517241         91  
1     -     0       2.517241        102  
2     -     0       2.517241        135  
3     m     0       2.517241         68  
4     m     0       2.517241         75  
5     m     0       2.517241         53  
6   m/e     0  yellow: icing         69  
7   m/e     0  yellow: icing         68  
8   m/e     0  yellow: icing         63  
9     m     0       2.517241         61  
10    m     0       2.517241         83  
11    -     0       2.517241         60  
12    -     0       2.517241         50  
13    -     0       2.517241         92  
14    -     0       2.517241        217  
15    -     0       2.517241        151  
16    -     0       2.517241         95  
17    -     0       2.517241        186  
18    -     1       2.517241        143  
19  m/e     1       2.517241         74  
20  m/e     1  yellow: storm        113  
21  m/e     4     red: storm         64  
22  m/e     4     red: storm         83  
23    e     1       2.517241        101  
24    e     0       2.517241         74  
25  m/e     0       2.517241         47  
26  m/e     0       2.517241         68  
27    -     0       2.517241         38  
28    -     0       2.517241        124  
29    m     0       2.517241         77  
30    m     0       2.517241         86

**Exercise 29:** the 'fog' column indicates whether there was fog in the morning ('m'), evening ('e') or both morning and evening ('m/e'). Add a column to `df` called 'fog_morning' of boolean values, indicating whether there was fog in the morning. Likewise, create a column 'fog_evening'. Finally, remove the original 'fog' column from `df`.

In [74]:
df["fog_morning"] = df["fog"].str.replace("m", "1").replace("", "0").replace("e", "0").replace("m/e", "1").astype(bool)
df["fog_evening"] = df["fog"].str.replace("e", "1").replace("", "0").replace("m", "0").replace("m/e", "1").astype(bool)
df.drop(columns=["fog"], inplace=True)

**Exercise 30:** sometimes it is useful to bin quantitative data into categories. Add a column to `df` called 'rain_intensity', whose values are strings representing rain categories depending on the amount of rain:
* 'no rain': 0 mm
* 'light rain': (0, 5] mm
* 'medium rain': (5, 10] mm
* 'heavy rain': (10, 50] mm

(Hint: pp. 203-205 of McKinney.)

In [78]:
df["rain_intensity"] = pd.cut(df["rain(mm)"], bins=[-1, 0, 5, 10, 50], labels=["no rain", "light rain", "medium rain", "heavy rain"])

**Exercise 31:** create a Series that counts the frequency of the occurrence of the different rain categories in January.

In [79]:
df["rain_intensity"].value_counts()

rain_intensity
light rain     16
no rain        10
medium rain     4
heavy rain      1
Name: count, dtype: int64

**Example:** the `sample` method draws a random sample of rows (or columns) from a Series or DataFrame. Drawing a sample of the size of the DataFrame is an easy way to randomly reorder the DataFrame.

In [80]:
df.sample(5)

0   index  YYYYMMDD  wind  T_ave  T_low  T_high  sun  rain_duration  rain(mm)  \
4       5  20180105   2.8    6.4    4.4     8.7  0.0            2.4       2.7   
27     23  20180128   3.4    9.7    6.9    10.8  1.7            0.0       0.1   
3      29  20180104   3.5    8.2    6.8    10.7  0.2            1.6       0.6   
15     13  20180116   3.2    5.2    2.5     6.9  2.9            3.9       8.3   
21     17  20180122   2.5    6.3    2.0     8.4  0.5            1.9       0.6   

0   snow weather alarm  accidents  fog_morning  fog_evening rain_intensity  
4      0      2.517241         75         True         True     light rain  
27     0      2.517241         38         True         True     light rain  
3      0      2.517241         68         True         True     light rain  
15     0      2.517241        151         True         True    medium rain  
21     4    red: storm         64         True         True     light rain

## Correlations

There are numerous modules that have built-in functions to compute correlations and covariances. They all have advantages and disadvantages and depending on the situation you can choose to use a specific module. We here give two examples: `corr` from the pandas module and `pearsonr` from SciPy's stats module.

**Example:** here we use pandas' `corr` function to compute the correlation between the 'wind' and 'accidents' columns. There are two disadvantages of the `corr` function: you need to make sure that the entries are all of the same type (e.g. by using `astype`) and it does not calculate the p-value of the correlation.

In [81]:
(df['wind'].astype(float)).corr(df['accidents'].astype(float))

np.float64(0.46885840130424355)

**Example:** the `pearsonr` function of the stats module of the SciPy package calculates both the correlation and the p-value (it returns a tuple of the two quantities). The main disadvantage of `pearsonr` is that it cannot handle missing values.

In [82]:
from scipy import stats # Importing the stats module of the SciPy package.

print(stats.pearsonr(df['wind'],df['accidents'])) # Pearsonr returns a tuple of the correlation and the p-value.

try:
    print(stats.pearsonr(df['rain_duration'],df['accidents']))
except:
    print("Pearsonr cannot handle missing values.")

PearsonRResult(statistic=np.float64(0.46885840130424355), pvalue=np.float64(0.007801325106885124))
PearsonRResult(statistic=np.float64(0.5922943256638151), pvalue=np.float64(0.0004475042650265651))


**Exercise 32:** define a function called `my_correlation` whose input arguments are two equal-length Series and whose output is a tuple of the correlation and the p-value. In contrast to the `pearsonr` function, `my_correlation` should be able to handle missing values by eliminating entry lines containing missing values from both(!) Series. Test `my_correlation` by computing the correlation and p-value of 'rain_duration' and 'accidents'.

In [98]:
def my_correlation(x: pd.Series, y: pd.Series) -> tuple[float, float]:
    x = x.dropna()
    y = y.dropna()
    return stats.pearsonr(x, y)
my_correlation(df['rain_duration'],df['accidents'])

PearsonRResult(statistic=np.float64(0.5922943256638151), pvalue=np.float64(0.0004475042650265651))

**Exercise 33:** create a dictionary whose keys are the column names and whose values are the correlation of the specific column with the 'accidents' column. Include only those columns for which a meaningful correlation can be computed. (Hint: define a list of names of the relevant columns and use a dictionary comprehension.)

In [95]:
df.head()

0  index  YYYYMMDD  wind  T_ave  T_low  T_high  sun  rain_duration  rain(mm)  \
0     11  20180101   3.2    6.8    5.2     8.8  2.1            6.9       4.7   
1      9  20180102   3.0    6.5    4.5     9.1  1.4            6.2       4.5   
2      8  20180103   4.8    8.8    5.4    11.2  0.2            5.1       9.5   
3     29  20180104   3.5    8.2    6.8    10.7  0.2            1.6       0.6   
4      5  20180105   2.8    6.4    4.4     8.7  0.0            2.4       2.7   

0  snow weather alarm  accidents  fog_morning  fog_evening rain_intensity  
0     0      2.517241         91         True         True     light rain  
1     0      2.517241        102         True         True     light rain  
2     0      2.517241        135         True         True    medium rain  
3     0      2.517241         68         True         True     light rain  
4     0      2.517241         75         True         True     light rain

In [99]:
dic = {c: my_correlation(df[c], df["accidents"]) for c in df.columns if c not in ["accidents", "YYYYMMDD", "index", "rain_intensity", "fog_morning", "fog_evening", "weather alarm"]}
dic

{'wind': PearsonRResult(statistic=np.float64(0.46885840130424355), pvalue=np.float64(0.007801325106885124)),
 'T_ave': PearsonRResult(statistic=np.float64(0.018689634590047174), pvalue=np.float64(0.92050965152688)),
 'T_low': PearsonRResult(statistic=np.float64(-0.1398856147873425), pvalue=np.float64(0.452926005345689)),
 'T_high': PearsonRResult(statistic=np.float64(0.2008689996727488), pvalue=np.float64(0.27857643770101326)),
 'sun': PearsonRResult(statistic=np.float64(-0.05774091178817214), pvalue=np.float64(0.7576760443316947)),
 'rain_duration': PearsonRResult(statistic=np.float64(0.5922943256638151), pvalue=np.float64(0.0004475042650265651)),
 'rain(mm)': PearsonRResult(statistic=np.float64(0.8131975402346316), pvalue=np.float64(2.7063566390647817e-08)),
 'snow': PearsonRResult(statistic=np.float64(-0.05551868991960596), pvalue=np.float64(0.7667406451549881))}

**Exercise 34:** create a dictionary similar to that of the previous exercise, but now only for correlations whose p-value is below 0.05. (Hint: apply a filtering using an `if` statement in the dictionary comprehension.)

In [101]:
filtered_dic = {k: v for k, v in dic.items() if v[1] < 0.05}
filtered_dic

{'wind': PearsonRResult(statistic=np.float64(0.46885840130424355), pvalue=np.float64(0.007801325106885124)),
 'rain_duration': PearsonRResult(statistic=np.float64(0.5922943256638151), pvalue=np.float64(0.0004475042650265651)),
 'rain(mm)': PearsonRResult(statistic=np.float64(0.8131975402346316), pvalue=np.float64(2.7063566390647817e-08))}

# 4. Data aggregation and group operations

Most of the material of this part is derived from sections 8.3, 10.1 and 10.2 of McKinney.

**Example:** creating a DataFrame whose row and column indices are both of MultiIndex type. Both indices have two levels. For exercises 1-6, always start working with the original DataFrame `frame` and do not redefine this variable!

In [102]:
frame = pd.DataFrame(np.random.randint(100, size=(4,5)),
                     index=[[2017, 2017, 2018, 2018], ['spring', 'fall', 'spring', 'fall']],
                     columns=[['Ohio', 'Ohio', 'Colorado', 'Alabama', 'Alabama'],
                              ['female', 'male', 'female', 'female', 'male']])

frame.index.names = ['year', 'semester']
frame.columns.names = ['state', 'gender']
frame

state           Ohio      Colorado Alabama     
gender        female male   female  female male
year semester                                  
2017 spring       73   18       10      56   57
     fall         45   75        8      63   74
2018 spring       62   75        5      72   79
     fall         13   86       53      27   89

**Example:** selecting the columns whose level-0 index is 'Ohio'.

In [103]:
frame['Ohio']

gender         female  male
year semester              
2017 spring        73    18
     fall          45    75
2018 spring        62    75
     fall          13    86

**Example:** selecting the entries associated with 'Ohio' (level 0) and 'male' (level 1).

In [104]:
frame['Ohio','male']

year  semester
2017  spring      18
      fall        75
2018  spring      75
      fall        86
Name: (Ohio, male), dtype: int32

**Exercise 35:** use `loc` to create a DataFrame with only the rows whose year is 2017.

In [107]:
frame.loc[2017]

state      Ohio      Colorado Alabama     
gender   female male   female  female male
semester                                  
spring       73   18       10      56   57
fall         45   75        8      63   74

**Exercise 36:** use `loc` to create a Series with only the row whose year is 2017 and semester is fall.

In [108]:
frame.loc[2017, "fall"]

state     gender
Ohio      female    45
          male      75
Colorado  female     8
Alabama   female    63
          male      74
Name: (2017, fall), dtype: int32

**Exercise 37:** use `iloc` to create a Series with only the row whose year is 2018 and semester is fall.

In [111]:
frame.iloc[3]

state     gender
Ohio      female    13
          male      86
Colorado  female    53
Alabama   female    27
          male      89
Name: (2018, fall), dtype: int32

**Example:** sorting the columns alphabetically for the 'state' part of the MultiIndex.

In [112]:
frame.sort_index(level='state', axis=1)

state         Alabama      Colorado   Ohio     
gender         female male   female female male
year semester                                  
2017 spring        56   57       10     73   18
     fall          63   74        8     45   75
2018 spring        72   79        5     62   75
     fall          27   89       53     13   86

**Example:** swapping the levels of the row MultiIndex.

In [113]:
frame.swaplevel('year', 'semester')

state           Ohio      Colorado Alabama     
gender        female male   female  female male
semester year                                  
spring   2017     73   18       10      56   57
fall     2017     45   75        8      63   74
spring   2018     62   75        5      72   79
fall     2018     13   86       53      27   89

**Exercise 38:** swap the levels of the column MultiIndex of `frame` and sort its rows by semester, in reverse alphabetical order. 

In [115]:
frame.sort_index(level='semester', ascending=False)

state           Ohio      Colorado Alabama     
gender        female male   female  female male
year semester                                  
2018 spring       62   75        5      72   79
2017 spring       73   18       10      56   57
2018 fall         13   86       53      27   89
2017 fall         45   75        8      63   74

**Example:** aggregating the data for the different states by summation.

In [116]:
frame.sum(level='gender', axis=1)

TypeError: sum() got an unexpected keyword argument 'level'

**Exercise 39:** calculate the mean quantities of `frame` for each combination of year and state.

In [124]:
fram

KeyError: 'state'

**Exercise 40:** use `stack` and `unstack` to turn the row index 'year' into a column index and the column index 'gender' into a row index.

**Example:** transforming the data of `frame` into a new `DataFrame` without MultiIndex. We call the new DataFrame `frame2`.

In [53]:
frame2 = frame.stack().stack().reset_index().rename({0 : 'quantity'}, axis=1)
frame2.quantity = frame2.quantity.astype(int)

frame2

**Example:** using `groupby` to compute the mean quantity per gender.

In [54]:
frame2.groupby('gender').mean().quantity

**Example:** we have already seen how to use `frame` to calculate the minimum quantity for each combination of year and state. This is done as follows.

In [55]:
frame.min(level='year').min(level='state', axis=1).sort_index(level=0).sort_index(axis=1)

**Exercise 41:** apply `groupby` to `frame2` to reproduce the DataFrame of the previous example.